### PR TITLE
Standard source incremental tests

### PR DESCRIPTION
**Pull request by sherifnada**
## What
* Test incremental syncs in standard source tests using a state file provided from the connector. 
* Connectors implementing standard source test now output a ConfiguredAirbyteCatalog instead of a Catalog. This is to solve the case of user-input cursors where there is no way to know the cursor. The exact contract is laid out in `getConfiguredCatalog`'s docstring.   

## How
* Standard tests now accept a state object output by the connector test suite. currently a source can output an empty state. 
* **Callout**: We run incremental tests if any streams in its discovered catalog claim to support incremental. This may be TooClever™. 
* Adds the following test cases to standard tests: 
  * Test that incremental sync produces records (we should ideally start testing exact record matches). Then, a second sync with the latest STATE message should produce no records. 
  * Test that an incremental sync with empty state produces records identical to a full refresh sync

## Reading order
* `TestSource.java`
* `standard_test.py` and `test_iface.py`
* anything else

## Follow up work 
- update all sources that support incremental to make use of this test suite. Most singer sources should automatically work with this.

### CODE REVIEW

**Review by jrhizor — APPROVED** [review]
I like it in general. 

I'm a bit concerned about how many redundant reads we're making to the external source. If we run into rate limiting problems we may have to do fewer redundant syncs (by memoizing `runRead` maybe?)

File: airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java [hunk 1]
```diff
@@ -101,12 +109,22 @@
   protected abstract JsonNode getConfig() throws Exception;
 
   /**
-   * Catalog to be used when attempting read operations.
+   * The catalog to use to validate the output of read operations. This will be used as follows:
+   * <p>
```
Comment (sherifnada) [inline airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java:113]: formatting borked this, will fix

File: airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java [hunk 2]
```diff
@@ -225,22 +243,100 @@ public void testRead() throws Exception {
...
+    ConfiguredAirbyteCatalog configuredCatalog = withFullRefreshSyncModes(getConfiguredCatalog());
+    final List<AirbyteRecordMessage> recordMessagesFirstRun = filterRecords(runRead(configuredCatalog));
+    final List<AirbyteRecordMessage> recordMessagesSecondRun = filterRecords(runRead(configuredCatalog));
     // the worker validates the messages, so we just validate the message, so we do not need to validate
     // again (as long as we use the worker, which we will not want to do long term).
     assertFalse(recordMessagesFirstRun.isEmpty());
     assertFalse(recordMessagesSecondRun.isEmpty());
-    assertSameMessages(recordMessagesSecondRun, recordMessagesSecondRun);
+
+    assertSameRecords(recordMessagesFirstRun, recordMessagesSecondRun);
```
Comment (sherifnada) [inline airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java:256]: bugfix, it was testing equality of the same list

Comment (cgardens) [reply airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java:256]: 🤦

File: airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java [hunk 3]
```diff
@@ -225,22 +243,100 @@ public void testRead() throws Exception {
...
+    ConfiguredAirbyteCatalog clone = Jsons.clone(catalog);
+    for (ConfiguredAirbyteStream configuredStream : clone.getStreams()) {
+      if (configuredStream.getStream().getSupportedSyncModes().contains(FULL_REFRESH)) {
+        configuredStream.setSyncMode(FULL_REFRESH);
+      }
+    }
+    return clone;
+  }
+
+  private boolean sourceSupportsIncremental() throws Exception {
```
Comment (cgardens) [inline airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java:332]: hahaha. TooClever™ indeed. i'm okay with this for now. @ChristopheDuong is adding a `supportIncremental` flag to the spec.json, so maybe we can use that instead when it is done.

**Review by cgardens — APPROVED** [review]

**Pull request merged**